### DISCUSSION TITLE
Discover & Discourage "Temporary Download Folder" on non-local drive?

### DISCUSSION THREAD

**Discussion by sanderjo** [sabnzbd/sabnzbd#2867] [Ideas]
What do we think about this:

Discover & Discourage "Temporary Download Folder" (aka Incomplete) on non-local drive?

With non-local drive:

- remote / mounted drive 
- docker: on the host, instead of within the docker filesystem

Reason: SABnzbd does a lot of writing & reading of small files to "Temporary Download Folder". Experience shows you can get weird behaviour, delays and errors. Using a local drive makes those problems dissappear.

Implementing seems not too difficult on Linux (and thus docker): in two steps (df and mount) you can see what kind of filesystem a directory is on.

```
$ df -h /run/user/1000/gvfs/asdf/asdf/asdf/
Filesystem      Size  Used Avail Use% Mounted on
gvfsd-fuse      1,8T  1,7T  174G  91% /run/user/1000/gvfs

$ mount | grep "/run/user/1000/gvfs" | awk '{ print $5 }'
fuse.gvfsd-fuse
```

So: remote mount. Not good for "Temporary Download Folder" 

```
$ df -h /home/sander/Downloads/
Filesystem      Size  Used Avail Use% Mounted on
/dev/nvme0n1p2  234G  126G   97G  57% /

$ mount | grep /dev/nvme0n1p2 | awk '{ print $5 }' | sort -u
ext4

```

ext4, so local, so good for "Temporary Download Folder"

**Comment by thezoggy** [comment]
in sab support you see all the time people have bad setups..  

- a 'local' drive, but badly suited for an incomplete folder.. like external usb drive directly attached.
- horribly mounted things like windows filesystem via docker on windows (so windows share -> linux virtual)
- truenas scale and them not able to do local drives and have to do nfs/iscsi shares..
- unraid users going through user share (not exclusive share or direct cache mount), so being hit with fuse overhead
- people using union file system (such as: mergerfs/mhddfs/unionfs/aufs/fuse) all have some overhead, some setup to do remote cloud storage.. so files are actually remote just appear local

Now with all that said, you cant really just guesstimate correctly base on the filesystem because it may not matter because of how its being used. Like inside the docker you might not have a clue how its being mounted on the host os.. to know its not optimal. Or someone using mergerfs and you just dont know how they have it configured to really make a good decision if you would want to target it or not. 

For example with docker volumes, people will use multiple mounts and not understand that to copy a file from one mount to another..this is across filesystem boundaries even if outside the docker it could be really on the same drive/filesystem. Thus you end up with a slow copy rather than a quick atomic move (pointer update).

I've seen people do a temp filesystem (ram drive) and then system doesnt have swap.. thus bad things happen when their ram drive gets filled up. So would you just flat out want to avoid ramdrives? i wouldnt because there are valid reasons for it, just people need to set it up correctly. 

So to avoid false positive of wrong assumptions it might be better to just warn users so we can educate them.

Maybe we should prob do a proper 'health/diag' page in sab where we could run through various checks to warn users if something might not be optimal from what we can check. Suggest things/link to pages about things we find.

**Comment by sanderjo** [comment]
Thanks.

I think we can detect most of those cases. It doesn't have to be perfect: we can start with easy cases, which will tackle 90% of the rainy day cases.

Incomplete on USB is not detectable, but not bad either, in my experience. And we have the current Wrench disk speed test for that. And if the Wrench finds a much faster disk, it could advice to use that. Hmmmm ... interesting ... SAB / Wrench could test home dir or temp to see the speed. If it's much better, advice that ... ?

Unraid: no idea what is going there (CPU and disk limitations? strange sas tooth results), but with any docker we can detect whether Incomplete is local (`overlay`) which is good, or something else (bad)

**Reply by thezoggy** [reply]
With unraid its the fact that use fuse shfs to merge the cache pool with the array pool to appear as one. that overlay causes a bit of io wait. fuse or any union fs come with a performance overhead. To sidestep this from happening you basically bypass fuse from being used. 

As people do not realize that when they go use /mnt/user/<usershare> its going to have the fuse overhead**, even if their share is set to use cache pool (or even only the cache). 
** In 6.12 they added 'exclusive share' to try and help people fix this by actually making a cache only share bypass using the fuse overhead (as it just makes a symlink `/mnt/user/<usershare>/` to `/mnt/cache/<usershare>`).

So to recap, there are two ways to achieve the same thing, just one uses a exclusive share while the other just is a direct path to it.

Before 6.12 you had to use direct path to the cache pool, so `/mnt/cache/.incomplete` for example. (not /mnt/user/*)
With 6.12+ you can keep the user share and make it 'exclusive', which bypasses fuse as its really just using the symlink to the direct path to the cache like above... 

for 'exclusive shares' to work you need to have it enabled in the global shares setting:
![firefox_2024-05-28_12-36-06](https://github.com/sabnzbd/sabnzbd/assets/500882/e6c64cd9-a533-488b-8b3c-aeb452a4caed)

example of exclusive share:
![firefox_2024-05-28_12-34-02](https://github.com/sabnzbd/sabnzbd/assets/500882/229bc5ea-fd03-4d97-a4d9-c92adfba4539)

otherwise, pre 6.12 (or you just dont want to do a share at all to reduce clutter), you can just set your volume mount to go directly to the cache like:
![firefox_2024-05-28_12-41-51](https://github.com/sabnzbd/sabnzbd/assets/500882/43d21a49-1ac4-4e50-b218-46731113cdcd)

Now why I have `.incomplete` there is because when your not doing an exclusive share, you run the risk of the unraid mover trying to move files you might not want. Can leverage the fact that the unraid mover will ignore dot folders, so that way it keeps its hands off and wont get touched.



The fuse overhead is not too terrible when your talking about a single file copy but io tanks when you have a bunch of little files like usenet downloading or torrenting. This is why when I have people do the download speed test before/after there is generally a ~50% difference in speed. This really rears its head when people have a large queue, or use direct unpack.. because then you high io wait and the backlog of stuff makes the system just struggle with it fighting for io. 

Then finally because people are bad about using paths they cause suboptimal workflow when having to copy between mounts which then adds even more io to it all. In sab/sonarr/radarr discord its a daily occurrence where we have to educate people on how to fix it, once we do its night and day difference.

**Reply by sanderjo** [reply]
Wow. I don't completely understand it, but I'll point Unraid users to your post when they report slowness / sawtooth speeds on their Unraid. #UnraidAlert

**Comment by dougie175** [comment]
I was referred here as I've followed a lot of the TRaSH Guides as they seem to do things right and I've managed to get atomic moves working perfectly which is great, but this has created a very unstable download speed to my SSD it constantly fluctuates from 70MB/s down to nothing, I believe this is because it's going through the Fuse system.

Reading above exclusive on the share would resolve this but when enabling exclusive it's not possible to have a secondary storage option and mover get everything downloaded during the day onto my array in the early hours. 

Is it possible to allow SABnzbd to write directly to the SSD without going through fuse, but then still get atmoic moves and mover still work in the early hours? i.e the best of all worlds?

The atmoic moves and mover action are more important to me then perfect continous download speeds so the way I have it at the moment is working in just trying to work it off it can be achieved? 

Thanks in advance

**Reply by thezoggy** [reply]
Yes, you would do the exclusive share for your incomplete folder so your bypassing fuse. Your complete folder would be to your user share (follow trash guides). Join the sab discord if you want to talk about it
